### PR TITLE
feat(journal): validate asqn on append

### LINKED ISSUE #6217: Journal should reject append if the sequence number is not greater than the previous record
#6216 introduces a journal module with new api. The new api includes an application sequence number which is a number provided by the application. The application can use either this sequence number or the index to seek to a record. The journal requires that the sequence number is always increasing. The first version implemented in #6216  does not validate this. However it is required that journal implementation rejects the append when the sequence number is not greater than the previous record.

Scope:
- Implement correct behaviour
- Introduce a new exception/error to indicate the failure if necessary
- Add tests

### PR DESCRIPTION
**Pull request by megglos**
## Description

Records to append are rejected in case the asqn is not bigger than the one of the previous entry.

## Related issues

closes #6217

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [x] The impact of the changes is verified by a benchmark
http://34.77.165.228/d/NzsO1mUnk/zeebe-overview?orgId=1&refresh=10s&from=now-3h&to=now&var-DS_PROMETHEUS=Prometheus&var-namespace=meg-6217-asqn-validate&var-pod=All&var-partition=All

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by deepthidevaki — CHANGES_REQUESTED** [review]
Thanks @megglos. Following scenarios also must be handled:

1. append(1, data); append(ASQN_IGNORE, data); append(1,data). This should fail.
2. When appending the first record in a new segment, the check for asqn should compare against the last asqn from the previous segment. You can write test for this in `SegmentedJournalTest` as you can control when a new segment is created.
3. After a restart, the check should compare against the last asqn written before restart. This is probably already handled. But we need a test for this scenario.

Let me know if you would like to discuss this.

**Comment by megglos** [comment]
> 1. append(1, data); append(ASQN_IGNORE, data); append(1,data). This should fail.

it's thus not allowed to have intermediate `ASQN_IGNORE` entries? let's best sync on improving my understanding there once you have time

**Comment by megglos** [comment]
> 1. append(1, data); append(ASQN_IGNORE, data); append(1,data). This should fail.

`JournalTest.shouldNotAppendRecordWithTooLowASqnIfPreviousRecordIsIgnoreASqn`

> 2. When appending the first record in a new segment, the check for asqn should compare against the last asqn from the previous segment. You can write test for this in `SegmentedJournalTest` as you can control when a new segment is created.
I went for two new test cases due to the fact that if the previous entry is an application entry, the validation will already prevent the creation of a new segment:
`shouldValidateAsqnBeforeCreatingNewSegment`
And a second to check that the last asqn is given to the new segment writer in case the last entry written is a ASQN_IGNORE entry.
`shouldValidateAsqnWhenWritingToNewSegment`

> 3. After a restart, the check should compare against the last asqn written before restart. This is probably already handled. But we need a test for this scenario.

On same segment
`shouldValidateAsqnWhenWritingAfterRestart`
On new segement with only an ASQN_IGNORE entry
`shouldValidateAsqnWhenWritingAfterRestartOnNewSegmentWithOnlyAsqnIgnoreRecord`

File: journal/src/test/java/io/camunda/zeebe/journal/file/SegmentedJournalTest.java
Comment (deepthidevaki) [inline journal/src/test/java/io/camunda/zeebe/journal/file/SegmentedJournalTest.java:725]: :wrench: This may be unnecessary.

**Review by deepthidevaki — APPROVED** [review]
:tada: Just one minor issue. No need for another review.

:x: `RandomizedRaftTest` is not using valid asqns. The test is not failing, but it is not doing anything useful because it cannot append any application entry to the journal. Please fix it before merge. You have to update `ControllableRaftContexts::clientAppend` such that it increases the asqn for each append.
```
      leaderRole.appendEntry(asqn, asqn, data, mock(AppendListener.class));   
 ```

File: journal/src/test/java/io/camunda/zeebe/journal/file/SegmentedJournalTest.java
Comment (megglos) [reply journal/src/test/java/io/camunda/zeebe/journal/file/SegmentedJournalTest.java:725]: it's here because of the following lambda in `assertThatThrownBy` that requires a final reference. However I went for just introducing:
```
    final SegmentedJournal reopenedJournal = openJournal(entriesPerSegment);
```

in the when blog when restarting.

**Review by deepthidevaki — APPROVED** [review]
Great! :100:  for improving the test.

**Comment by deepthidevaki** [comment]
bors merge

**Pull request merged**
